### PR TITLE
Fix OPTEE memory leak issue 

### PR DESCRIPTION
**Pull request by Summer-ARM**
It's better to call crypto_xxx_free_ctx instead of simple free
for memory allocated by crypto_xxx_alloc_ctx.

Signed-off-by: Summer Qin <summer.qin@arm.com>
#2112

### CODE REVIEW

File: core/arch/arm/kernel/ree_fs_ta.c [hunk 1]
```diff
@@ -125,7 +125,7 @@ static TEE_Result ta_open(const TEE_UUID *uuid,
 	hash_algo = TEE_DIGEST_HASH_TO_ALGO(shdr->algo);
 	res = crypto_hash_alloc_ctx(&hash_ctx, hash_algo);
 	if (res != TEE_SUCCESS)
-		goto error_free_payload;
+		goto error_free_hash;
```
Comment (jforissier) [inline core/arch/arm/kernel/ree_fs_ta.c]: No. When `crypto_hash_alloc_ctx()` returns an error code, you can't rely on the value of the context. It is the job of the alloc function to clean up after itself should any error occur.

File: core/arch/arm/pta/secstor_ta_mgmt.c [hunk 2]
```diff
@@ -70,7 +70,7 @@ static TEE_Result install_ta(struct shdr *shdr, const uint8_t *nw,
 	hash_algo = TEE_DIGEST_HASH_TO_ALGO(shdr->algo);
 	res = crypto_hash_alloc_ctx(&hash_ctx, hash_algo);
 	if (res)
-		goto err;
+		goto err_free_hash_ctx;
```
Comment (jforissier) [inline core/arch/arm/pta/secstor_ta_mgmt.c]: Same here.

File: core/tee/fs_htree.c [hunk 3]
```diff
@@ -462,8 +462,10 @@ static TEE_Result authenc_init(void **ctx_ret, TEE_OperationMode mode,
 	}
 
 	res = crypto_authenc_alloc_ctx(&ctx, alg);
-	if (res != TEE_SUCCESS)
+	if (res != TEE_SUCCESS) {
+		crypto_authenc_free_ctx(ctx, alg);
```
Comment (jforissier) [inline core/tee/fs_htree.c]: No.

File: core/tee/fs_htree.c [hunk 4]
```diff
@@ -495,11 +497,12 @@ static TEE_Result authenc_init(void **ctx_ret, TEE_OperationMode mode,
...
 
 exit:
-	if (res == TEE_SUCCESS)
+	if (res == TEE_SUCCESS) {
 		*ctx_ret = ctx;
-	else
+	} else {
 		crypto_authenc_final(ctx, alg);
-
+		crypto_authenc_free_ctx(ctx, alg);
```
Comment (jforissier) [inline core/tee/fs_htree.c]: `crypto_authenc_final()` should be done only if `crypto_authenc_init()` succeeded.
`crypto_authenc_free_ctx()` should be called only if `crypto_authenc_alloc()` succeeded (that is the actual fix for the memory leak)
These two changes should go in there own patch "core: FS: htree: fix potential memory leak".

File: core/tee/fs_htree.c [hunk 5]
```diff
@@ -584,13 +587,12 @@ static TEE_Result verify_node(struct traverse_arg *targ,
...
-	void *ctx;
+	void *ctx = NULL;
 
 	res = crypto_hash_alloc_ctx(&ctx, TEE_FS_HTREE_HASH_ALG);
-	if (res != TEE_SUCCESS)
-		return res;
+	if (res == TEE_SUCCESS)
+		res = htree_traverse_post_order(ht, verify_node, ctx);
 
-	res = htree_traverse_post_order(ht, verify_node, ctx);
```
Comment (jforissier) [inline core/tee/fs_htree.c]: No. Same as above.

File: core/tee/fs_htree.c [hunk 6]
```diff
@@ -599,17 +601,15 @@ static TEE_Result verify_tree(struct tee_fs_htree *ht)
...
-	ht->root.dirty = true;
-
-	res = calc_node_hash(&ht->root, &ht->imeta.meta, ctx,
-			     ht->root.node.hash);
+	if (res == TEE_SUCCESS) {
+		ht->root.id = 1;
+		ht->root.dirty = true;
+		res = calc_node_hash(&ht->root, &ht->imeta.meta, ctx,
+				     ht->root.node.hash);
+	}
```
Comment (jforissier) [inline core/tee/fs_htree.c]: No.

File: core/tee/fs_htree.c [hunk 7]
```diff
@@ -773,7 +773,7 @@ TEE_Result tee_fs_htree_sync_to_storage(struct tee_fs_htree **ht_arg,
 
 	res = crypto_hash_alloc_ctx(&ctx, TEE_FS_HTREE_HASH_ALG);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
```
Comment (jforissier) [inline core/tee/fs_htree.c]: No.

File: core/tee/tadb.c [hunk 8]
```diff
@@ -284,12 +284,14 @@ static TEE_Result tadb_authenc_init(TEE_OperationMode mode,
...
-	void *ctx;
+	void *ctx = NULL;
 	const size_t enc_size = entry->prop.custom_size + entry->prop.bin_size;
 
 	res = crypto_authenc_alloc_ctx(&ctx, TADB_AUTH_ENC_ALG);
-	if (res)
+	if (res) {
+		crypto_authenc_free_ctx(ctx, TADB_AUTH_ENC_ALG);
 		return res;
+	}
```
Comment (jforissier) [inline core/tee/tadb.c]: No.

File: core/tee/tee_cryp_concat_kdf.c [hunk 9]
```diff
@@ -26,7 +26,7 @@ TEE_Result tee_cryp_concat_kdf(uint32_t hash_id, const uint8_t *shared_secret,
 
 	res = crypto_hash_alloc_ctx(&ctx, hash_algo);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
```
Comment (jforissier) [inline core/tee/tee_cryp_concat_kdf.c]: No.

File: core/tee/tee_cryp_hkdf.c [hunk 10]
```diff
@@ -137,7 +137,7 @@ static TEE_Result hkdf_expand(uint32_t hash_id, const uint8_t *prk,
 	}
 
 out:
-	free(ctx);
+	crypto_mac_free_ctx(ctx, hmac_algo);
```
Comment (jforissier) [inline core/tee/tee_cryp_hkdf.c]: OK, this can go in a patch with subject: "Call `crypto_mac_free_ctx()` instead of `free()`"

File: core/tee/tee_cryp_pbkdf2.c [hunk 11]
```diff
@@ -90,7 +90,7 @@ TEE_Result tee_cryp_pbkdf2(uint32_t hash_id, const uint8_t *password,
 
 	res = crypto_mac_alloc_ctx(&hmac_parms.ctx, hmac_parms.algo);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
```
Comment (jforissier) [inline core/tee/tee_cryp_pbkdf2.c]: No.

File: core/tee/tee_cryp_utl.c [hunk 12]
```diff
@@ -76,7 +76,7 @@ TEE_Result tee_hash_createdigest(uint32_t algo, const uint8_t *data,
 
 	res = crypto_hash_alloc_ctx(&ctx, algo);
 	if (res)
-		return res;
+		goto out;
```
Comment (jforissier) [inline core/tee/tee_cryp_utl.c]: No.

File: core/tee/tee_fs_key_manager.c [hunk 13]
```diff
@@ -49,7 +49,7 @@ static TEE_Result do_hmac(void *out_key, size_t out_key_size,
 
 	res = crypto_mac_alloc_ctx(&ctx, TEE_FS_KM_HMAC_ALG);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto exit;
```
Comment (jforissier) [inline core/tee/tee_fs_key_manager.c]: No.

File: core/tee/tee_fs_key_manager.c [hunk 14]
```diff
@@ -108,7 +108,7 @@ TEE_Result tee_fs_fek_crypt(const TEE_UUID *uuid, TEE_OperationMode mode,
 
 	res = crypto_cipher_alloc_ctx(&ctx, TEE_FS_KM_ENC_FEK_ALG);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto exit;
```
Comment (jforissier) [inline core/tee/tee_fs_key_manager.c]: No.

File: core/tee/tee_fs_key_manager.c [hunk 15]
```diff
@@ -196,7 +196,7 @@ static TEE_Result aes_ecb(uint8_t out[TEE_AES_BLOCK_SIZE],
 
 	res = crypto_cipher_alloc_ctx(&ctx, algo);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
```
Comment (jforissier) [inline core/tee/tee_fs_key_manager.c]: No.

File: core/tee/tee_fs_key_manager.c [hunk 16]
```diff
@@ -263,7 +263,7 @@ TEE_Result tee_fs_crypt_block(const TEE_UUID *uuid, uint8_t *out,
 	/* Run AES CBC */
 	res = crypto_cipher_alloc_ctx(&ctx, algo);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto exit;
```
Comment (jforissier) [inline core/tee/tee_fs_key_manager.c]: No.

File: core/tee/tee_rpmb_fs.c [hunk 17]
```diff
@@ -410,7 +410,7 @@ static TEE_Result tee_rpmb_mac_calc(uint8_t *mac, uint32_t macsize,
 
 	res = crypto_mac_alloc_ctx(&ctx, TEE_ALG_HMAC_SHA256);
 	if (res)
-		return res;
+		goto func_exit;
```
Comment (jforissier) [inline core/tee/tee_rpmb_fs.c]: No.

File: core/tee/tee_cryp_hkdf.c (hunk 10 above)
Comment (Summer-ARM) [reply core/tee/tee_cryp_hkdf.c]: Hi @jforissier 
Thanks for your nice explanation.
As you said before, crypto_authenc_free_ctx() should be called only if crypto_authenc_alloc() succeeded. But I see in core/tee/tee_cryp_hkdf.c, it goto out when alloc fails and even before alloc.
https://github.com/OP-TEE/optee_os/blob/b924c494920f9197b4b6c9fc313b174306282f7b/core/tee/tee_cryp_hkdf.c#L76-L90

Comment (jenswi-linaro) [reply core/tee/tee_cryp_hkdf.c]: This is sadly an error introduced when the crypto API was changed.

Comment (jforissier) [reply core/tee/tee_cryp_hkdf.c]: Yeah `free()` should be `crypto_mac_free_ctx()` but please note that `free(NULL)` and similarly `crypto_mac_free_ctx(NULL, algo)` *are* valid.

Comment (jenswi-linaro) [reply core/tee/tee_cryp_hkdf.c]: Hrm, @jforissier not entirely true, see:
https://github.com/OP-TEE/optee_os/blob/master/core/crypto/crypto.c#L103-L106

Comment (Summer-ARM) [reply core/tee/tee_cryp_hkdf.c]: Ok, I will update my patch after you send a PR

Comment (jforissier) [reply core/tee/tee_cryp_hkdf.c]: https://github.com/OP-TEE/optee_os/pull/2134

**Comment by jforissier** [comment]
https://github.com/OP-TEE/optee_os/pull/2145

**Comment by jforissier** [comment]
@Summer-ARM thanks for the update. I have reworked the code slightly in a new PR so I'm closing this one if you don't mind.
